drivers/input/input_pinnacle.c

```diff
@@ -0,0 +1,627 @@
+#define PINNACLE_SYS_CONFIG1_SHUTDOWN       BIT(1)
+#define PINNACLE_SYS_CONFIG1_LOW_POWER_MODE BIT(2)
+
+/* FeedConfig1 definition */
+#define PINNACLE_FEED_CONFIG1_FEED_ENABLE        BIT(0)
+#define PINNACLE_FEED_CONFIG1_DATA_MODE_ABSOLUTE BIT(1)
+#define PINNACLE_FEED_CONFIG1_FILTER_DISABLE     BIT(2)
+#define PINNACLE_FEED_CONFIG1_X_DISABLE          BIT(3)
+#define PINNACLE_FEED_CONFIG1_Y_DISABLE          BIT(4)
+/* X max to 0 */
+#define PINNACLE_FEED_CONFIG1_X_DATA_INVERT      BIT(6)
+/* Y max to 0 */
+#define PINNACLE_FEED_CONFIG1_Y_DATA_INVERT      BIT(7)
+
+/* FeedConfig2 definition */
+#define PINNACLE_FEED_CONFIG2_INTELLIMOUSE_ENABLE   BIT(0)
+#define PINNACLE_FEED_CONFIG2_ALL_TAPS_DISABLE      BIT(1)
+#define PINNACLE_FEED_CONFIG2_SECONDARY_TAP_DISABLE BIT(2)
+#define PINNACLE_FEED_CONFIG2_SCROLL_DISABLE        BIT(3)
+#define PINNACLE_FEED_CONFIG2_GLIDE_EXTEND_DISABLE  BIT(4)
+/* 90 degrees rotation */
+#define PINNACLE_FEED_CONFIG2_SWAP_X_AND_Y          BIT(7)
+
+/* Special definitions */
+#define PINNACLE_FB (0xFB) /* Filler byte */
+#define PINNACLE_FC (0xFC) /* Auto-increment byte */
+
+/* Read and write masks */
+#define PINNACLE_READ_MSK  (0xA0)
+#define PINNACLE_WRITE_MSK (0x80)
+
+/* Read and write register addresses */
+#define PINNACLE_READ_REG(addr)  (PINNACLE_READ_MSK | addr)
+#define PINNACLE_WRITE_REG(addr) (PINNACLE_WRITE_MSK | addr)
+
+struct pinnacle_active_range {
+	uint16_t x_min;
+	uint16_t x_max;
+	uint16_t y_min;
+	uint16_t y_max;
+};
+
+struct pinnacle_resolution {
+	uint16_t x;
+	uint16_t y;
+};
+
+struct pinnacle_bus;
+
+typedef bool (*pinnacle_bus_is_ready)(const struct pinnacle_bus *bus);
+typedef int (*pinnacle_bus_write)(const struct pinnacle_bus *bus, uint8_t address, uint8_t value);
+typedef int (*pinnacle_bus_read)(const struct pinnacle_bus *bus, uint8_t address, uint8_t *value);
+typedef int (*pinnacle_bus_seq_read)(const struct pinnacle_bus *bus, uint8_t address, uint8_t *data,
+				     uint8_t count);
+
+struct pinnacle_bus {
+	union {
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
+		struct i2c_dt_spec i2c;
+#endif
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+		struct spi_dt_spec spi;
+	};
+#endif
+	pinnacle_bus_is_ready is_ready;
+	pinnacle_bus_write write;
+	pinnacle_bus_read read;
+	pinnacle_bus_seq_read seq_read;
+};
+
+struct pinnacle_config {
+	const struct pinnacle_bus bus;
+	struct gpio_dt_spec dr_gpio;
+	uint8_t idle_packets_count;
+
+	bool clipping_enabled;
+	bool scaling_enabled;
+
+	struct pinnacle_active_range active_range;
+	struct pinnacle_resolution resolution;
+};
+
+struct pinnacle_sample {
+#if defined(CONFIG_INPUT_PINNACLE_DATA_MODE_ABSOLUTE)
+	uint16_t x;
+	uint16_t y;
+	uint8_t z;
+#elif defined(CONFIG_INPUT_PINNACLE_DATA_MODE_RELATIVE)
+	int16_t x;
+	int16_t y;
+#endif
+};
+
+struct pinnacle_data {
+	struct pinnacle_sample sample;
+	const struct device *dev;
+	struct gpio_callback dr_cb_data;
+	struct k_work work;
+};
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
```

Comment:
The I2C implementation is in progress.

Reply:
For some reason, I have an issue with communicating to Pinnacle through I2C. I have nRF52-DK where I connect Pinnacle to I2C0. I can see that both SDA and SLC are in active high state, which is correct. However, when a transmission starts, both lines goes down and the transmission fails. I'm trying to figure out what's happening wrong and why. While I'm trying to figure out what's the problem, I'll keep I2C function unimplemented.

Reply:
Do you have the pullups on the i2c lines? on nrf52dk you can enable the integrated one by shorting the "DETECT" line on the 3x2 connector to ground (you lose the LED and switches).

Regardless, feel free to implement them, I have one of these and I can test it for you.

Reply:
You was absolutely right that pull-up resistors were missing. I was reading the schematic incorrectly and thought that were connected when `DETECT` is not connected, which was wrong. Thank you very much for this advice!